com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/TaskBuilder.java

```diff
@@ -0,0 +1,362 @@
+        TimeProfiler.stop();
+        return taskResult;
+    }
+
+
+    public Set<IResource> getAllOutputs() {
+        return allOutputs;
+    }
+
+
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    public List<TaskResult> build(IProgress monitor) throws IOException, CompileExceptionError {
+        TimeProfiler.start("Build tasks");
+        logger.info("Build tasks");
+        long tstart = System.currentTimeMillis();
+
+        buildContainsChanges = false;
+        boolean abort = false;
+        int maxConcurrentHighMemoryTasks = this.nThreads;
+        List<Callable<TaskResult>> tasksToSubmit = new ArrayList<>();
+        Map<String, Integer> taskNameCounter = new HashMap<>();
+        while (!tasks.isEmpty() && !abort) {
+
+            // create a list of tasks to build this iteration
+            // - ignore tasks with unresolved dependencies
+            // - build game.project last as the only remaining task (see #9553)
+            // - never build more than one material at a time since there is
+            //   some kind of threading issue with the command line tools when
+            //   building two materials which share a program 
+            // - restrict atlas/tileset builders per iteration (for memory reasons)
+            int remainingTasksCount = tasks.size();
+            tasksToSubmit.clear();
+            taskNameCounter.clear();
+            Set<String> taskNames = new HashSet<>();
+            for (Task task : tasks) {
+                if (task.getBuilder().isGameProjectBuilder() && remainingTasksCount > 1) continue;
+                String taskName = task.getName();
+                if (taskName.equals("TileSet")) taskName = "Atlas";
+                int count = taskNameCounter.getOrDefault(taskName, 0);
+                if (taskName.equals("Atlas") && (count == maxConcurrentHighMemoryTasks)) continue;
+                if (taskName.equals("Material") && count == 1) continue;
+                if (hasUnresolvedDependencies(task)) continue;
+                tasksToSubmit.add(createCallableTask(task, monitor));
+                taskNameCounter.put(taskName, count + 1);
+            }
+
+            try {
+                boolean retryAnyTask = false;
+                List<Future<TaskResult>> futures = this.executorService.invokeAll(tasksToSubmit);
+                for (Future<TaskResult> future : futures) {
+                    TaskResult result = future.get();
+                    // should the task be retried again?
+                    // this can happen if running out of memory while building
+                    if (result.getResult() == Result.RETRY) {
+                        retryAnyTask = true;
+                        if (maxConcurrentHighMemoryTasks > 1) {
+                            maxConcurrentHighMemoryTasks--;
+                            logger.info("Task must be retried. Reducing number of concurrent high memory tasks to %d", maxConcurrentHighMemoryTasks);
```

Comment:
will it be in a regular log info? it's a very useful for CI configuration (e.g. increase memory per thread)

Reply:
I'll change it to a warning